### ISSUE TITLE
Segfault when trying to run tests locally

### ISSUE THREAD

**Issue by goerz** [qutip/qutip#1093]
I'm trying to run qutip's test suite locally on my Macbook.

From a clean checkout of the `master` branch, I'm doing the following:

~~~
conda create -q -n qutip_testing_env -y python=3.7
conda install -n qutip_testing_env -y blas=*=openblas numpy scipy nose cython ipython pytest
mkdir qutip_testing
cd qutip_testing
rm -rf /Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/lib/python3.7/site-packages/qutip*
rm -rf ../build ../dist/ ../qutip.egg-info/
(cd ../ && /Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/bin/python setup.py install)
/Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/bin/python -m qutip.about
/Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/bin/nosetests --verbosity=2 --with-coverage --cover-package=qutip qutip
~~~

The resulting output is:
~~~
...
QuTiP: Quantum Toolbox in Python
================================
Copyright (c) QuTiP team 2011 and later.
Original developers: R. J. Johansson & P. D. Nation.
Current admin team: Alexander Pitchford, Paul D. Nation, Nathan Shammah, Shahnawaz Ahmed, Neill Lambert, and Eric Giguère.
Project Manager: Franco Nori.
Currently developed through wide collaboration. See https://github.com/qutip for details.

QuTiP Version:      4.5.0.dev0+d377c997
Numpy Version:      1.17.2
Scipy Version:      1.3.1
Cython Version:     0.29.13
Matplotlib Version: None
Python Version:     3.7.4
Number of CPUs:     2
BLAS Info:          OPENBLAS
OPENMP Installed:   False
INTEL MKL Ext:      False
Platform Info:      Darwin (x86_64)
Installation path:  /Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/lib/python3.7/site-packages/qutip-4.5.0.dev0+
d377c997-py3.7-macosx-10.9-x86_64.egg/qutip
==============================================================================
Please cite QuTiP in your publication.
==============================================================================
For your convenience a bibtex reference can be easily generated using `qutip.cite()`
nose.plugins.cover: ERROR: Coverage not available: unable to import coverage module
/Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/lib/python3.7/site-packages/qutip-4.5.0.dev0+d377c997-py3.7-macos
x-10.9-x86_64.egg/qutip/__init__.py:161: UserWarning: matplotlib not found: Graphics will not work.
  warnings.warn("matplotlib not found: Graphics will not work.")
Transform 2-level to eigenbasis and back ... ok
Transform 10-level real-values to eigenbasis and back ... ok
Transform 10-level to eigenbasis and back ... ok
Transform 10-level imag to eigenbasis and back ... ok
Consistency between transformations of kets and density matrices ... ok
Check diagonalization via eigenbasis transformation ... ok
Check Qobj eigs and direct eig solver transformations match ... ok
Check Qobj eigs and direct eig solver reverse transformations match ... ok
brmesolve: simple qubit ... ok
brmesolve: c_ops alone ... ok
brmesolve: c_ops with a_ops ... ok
brmesolve: harmonic oscillator, zero temperature ... ok
brmesolve: harmonic oscillator, finite temperature ... ok
brmesolve: harmonic oscillator, finite temperature, states ... ok
brmesolve: Jaynes-Cummings model, zero temperature ... ok
brmesolve: Check for #572 bug. ... ok
brmesolve: input list of Qobj ... ok
td_brmesolve: passes all brmesolve tests ... ok
td_brmesolve: time-dependent a_ops ... ok
td_brmesolve: time-dependent a_ops tuple of strings ... ok
td_brmesolve: time-dependent a_ops tuple interp ... ok                                                                                 td_brmesolve: time-dependent a_ops & c_ops interp ... ok
td_brmesolve: non-Hermitian e_ops check ... ok
td_brmesolve: states check ... ok
td_brmesolve: split ops #1 ... ok
td_brmesolve: split ops #2 ... ok
td_brmesolve: split ops, Cubic_Spline td-terms ... ok
td_brmesolve: split ops, multiple ... ok
td_brmesolve: Hamiltonian args ... ok
BR Tools : zheevr ... Segmentation fault: 11
~~~

**Comment by tehruhn** [comment]
Hi @goerz ,

I tried to replicate your issue on my Linux machine, and it seems to work fine :
`(qutip-env) tarun@trolldem0rt:~/riken-work/qutip/qutip$ cd ~/miniconda3/envs/qutip-env/bin/
(qutip-env) tarun@trolldem0rt:~/miniconda3/envs/qutip-env/bin$ nosetests --verbosity=2 --with-coverage --cover-package=qutip qutip
nose.plugins.cover: ERROR: Coverage not available: unable to import coverage module
Transform 2-level to eigenbasis and back ... ok
Transform 10-level real-values to eigenbasis and back ... ok
Transform 10-level to eigenbasis and back ... ok
Transform 10-level imag to eigenbasis and back ... ok
Consistency between transformations of kets and density matrices ... ok
Check diagonalization via eigenbasis transformation ... ok
Check Qobj eigs and direct eig solver transformations match ... ok
Check Qobj eigs and direct eig solver reverse transformations match ... ok
brmesolve: simple qubit ... ok
brmesolve: c_ops alone ... ok
brmesolve: c_ops with a_ops ... ok
brmesolve: harmonic oscillator, zero temperature ... ok
brmesolve: harmonic oscillator, finite temperature ... ok
brmesolve: harmonic oscillator, finite temperature, states ... ok
brmesolve: Jaynes-Cummings model, zero temperature ... ok
brmesolve: Check for #572 bug. ... ok
brmesolve: input list of Qobj ... ok
td_brmesolve: passes all brmesolve tests ... ok
td_brmesolve: time-dependent a_ops ... ok
td_brmesolve: time-dependent a_ops tuple of strings ... ok
td_brmesolve: time-dependent a_ops tuple interp ... ok
td_brmesolve: time-dependent a_ops & c_ops interp ... ok
td_brmesolve: non-Hermitian e_ops check ... ok
td_brmesolve: states check ... ok
td_brmesolve: split ops #1 ... ok
td_brmesolve: split ops #2 ... ok
td_brmesolve: split ops, Cubic_Spline td-terms ... ok
td_brmesolve: split ops, multiple ... ok
td_brmesolve: Hamiltonian args ... ok
BR Tools : zheevr ... ok
BR Tools : dense operator to eigenbasis ... ok
BR Tools : vector to eigenbasis ... ok
BR Tools : eigvector to fockbasis ... ok
BR Tools : vector roundtrip transform ... ok
BR Tools : Diagonal liouvillian mult ... ok
BR Tools : cop_super_mult ... ok
BR Tools : br_term_mult ... ok
control.pulseoptim: Hadamard gate with linear initial pulses ... ok
control.pulseoptim: Hadamard gate with linear initial pulses (no stats) ... ok
`

This shows that it is probably not an issue with the source code. I was looking at the output you provided from `qutip.about`, and you seem to be using `QuTiP Version:      4.5.0.dev0+d377c997`.

I suspect this could be the issue, since the dev version is probably not stable yet. So I'd recommend you try this with a stable release of QuTiP.
Would you try running the same with QuTiP 4.4.1, by explicitly installing the same in conda as follows, and get back to us? :
`conda install qutip=4.4.1`

**Comment by nathanshammah** [comment]
Let me just mention that there are ongoing discussions on switching to / partly introducing Numba.

**Comment by Ericgig** [comment]
Would it help to run just a part of the tests?
`nose.run(defaultTest="qutip.tests.test_mcsolve", argv=['nosetests', '-v'])`
That way you can go around the td_brmesolve's tests. 

Can anybody who has this error (@arnelg) and some time, check if:
1. You get the same error with 4.4.1, nobody touched the brmesolve for years as far as I know.
2. Only brmesolve fails, here is a script that will run all other tests, just set `path_to_qutip`
```
path_to_qutip = "/home/.../qutip"
import glob
testfiles = glob.glob(path_to_qutip+"/qutip/tests/*.py")
testfiles = [testfile for testfile in testfiles if "br" not in testfile] # skip brmesolve stuff
testfiles = [testfile for testfile in testfiles if "__init__" not in testfile] # remove __init__.py
tests = [testfile.split("/")[-1][:-3] for testfile in testfiles]
for test in tests:
    nose.run(defaultTest="qutip.tests." + test, argv=['nosetests', '-v'])
```

**Comment by goerz** [comment]
I had a bit more success by switching as much as possible in the setup of the testing environment to pip:
~~~
conda create -q -n qutip_testing_env -y python=3.7
conda install -n qutip_testing_env -y numpy scipy cython
/Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/bin/python -m pip install -e .
/Users/goerz/.pyenv/versions/miniconda3-latest/envs/qutip_testing_env/bin/python -m pip install nose ipython
~~~
I also added add `skipIf` decorator to `qutip/tests/test_stochastic_se.py:test_smesolve_homodyne_methods`, which is the only test that consistently fails on `master`.

With this, the test suite runs through about 50% of the time. The other 50% it crashes with the following:
~~~
...
PIQS: Test the warning for diagonal Hamiltonians to use internal solver. ... ok
Propagator: HO ('single mode') ... ok
Propagator: HO ('batch mode') ... ok
Propagator: HO parallel ... ok
Propagator: str td format ... ok
Propagator: func td format ... ok
Propagator: steady state ... python(52709,0x10fc1f5c0) malloc: Incorrect checksum for freed object 0x7fcbf338af10: probably modified after being freed.
Corrupt value: 0x3ff0000000000000
python(52709,0x10fc1f5c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
~~~

**Comment by nathanshammah** [comment]
Yes, I also get a few abort trap 6 crashes. I wonder why on Travis CI they seem to always run fine lately. By the way, we have migrated to `pytest`. Hopefully this will help.

**Comment by nathanshammah** [comment]
See #1160.

**Comment by jakelishman** [comment]
This particular segfault is solved by #1288 - it's related to a bust `zheevr` implementation in the Mac OpenBLAS stack somewhere or another (may even be in `scipy`'s binding, we never tracked it down exactly).  This doesn't mean all segfaults on Mac are fixed - see also #1160, which I believe still occasionally rears its head and we have no idea about.
